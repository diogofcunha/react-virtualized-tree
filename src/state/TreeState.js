import {getFlattenedTreePaths} from '../selectors/getFlattenedTree';
import {getNodeFromPath} from '../selectors/nodes';

export class State {
  flattenedTree = null;
  tree = null;

  constructor(tree, flattenedTree) {
    this.tree = tree;
    this.flattenedTree = flattenedTree || getFlattenedTreePaths(tree);
  }
}

export const validateState = state => {
  if (!(state instanceof State)) {
    throw new Error(`Expected a State instance but got ${typeof state}`);
  }
};

/**
 * Immutable structure that represents the TreeState.
 */
export default class TreeState {
  /**
   * Given a state, finds a node at a certain row index.
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @return {State} An internal state representation
   */
  static getNodeAt = (state, index) => {
    validateState(state);

    const rowPath = state.flattenedTree[index];

    if (!rowPath) {
      throw Error(
        `Tried to get node at row "${index}" but got nothing, the tree are ${state.flattenedTree.length} visible rows`,
      );
    }

    return getNodeFromPath(rowPath, state.tree);
  };

  /**
   * Given a state, finds a node deepness at a certain row index.
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @return {number} The node deepness
   */
  static getNodeDeepness = (state, index) => {
    validateState(state);

    const rowPath = state.flattenedTree[index];

    if (!rowPath) {
      throw Error(
        `Tried to get node at row "${index}" but got nothing, the tree are ${state.flattenedTree.length} visible rows`,
      );
    }

    return rowPath.length - 1;
  };

  /**
   * Given a state and an index, finds the number of visible descendants
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @return {number} The number of visible descendants
   */
  static getNumberOfVisibleDescendants = (state, index) => {
    const {id} = TreeState.getNodeAt(state, index);

    const {flattenedTree} = state;
    let i;

    for (i = index; i < flattenedTree.length; i++) {
      const path = flattenedTree[i];

      if (!path.some(p => p === id)) {
        break;
      }
    }

    return Math.max(i - 1 - index, 0);
  };

  /**
   * Given a state, gets the tree
   * @param {State} state - The current state
   * @return {Node[]} The tree
   */
  static getTree = state => {
    validateState(state);

    return state.tree;
  };

  /**
   * Creates an instance of state.
   * @param {Node[]} tree - The original tree
   * @return {State} An internal state representation
   */
  static createFromTree = tree => {
    if (!tree) {
      throw Error('A falsy tree was supplied in tree creation');
    }

    if (!Array.isArray(tree)) {
      throw Error('An invalid tree was supplied in creation');
    }

    return new State(tree);
  };
}
