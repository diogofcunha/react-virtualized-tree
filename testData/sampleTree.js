export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
