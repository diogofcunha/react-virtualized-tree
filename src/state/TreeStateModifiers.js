import {
  getFlattenedTreePaths,
  doesChangeAffectFlattenedTree,
  isNodeExpanded,
  nodeHasChildren,
} from '../selectors/getFlattenedTree';
import TreeState, {State} from './TreeState';
import {replaceNodeFromTree, deleteNodeFromTree} from '../selectors/nodes';

/**
 * @callback setNode
 * @param {Node} node - current node value
 * @return {Node} The updated node
 */

/**
 * Set of Tree State Modifiers
 */
export default class TreeStateModifiers {
  /**
   * Given a state, finds a node at a certain row index.
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @param {setNode|Node} nodeUpdate - A function to update the node
   * @return {State} An internal state representation
   */
  static editNodeAt = (state, index, nodeUpdate) => {
    const node = TreeState.getNodeAt(state, index);
    const updatedNode = typeof nodeUpdate === 'function' ? nodeUpdate(node) : nodeUpdate;
    const flattenedTree = [...state.flattenedTree];
    const flattenedNodeMap = flattenedTree[index];
    const parents = flattenedNodeMap.slice(0, flattenedNodeMap.length - 1);

    if (doesChangeAffectFlattenedTree(node, updatedNode)) {
      const numberOfVisibleDescendants = TreeState.getNumberOfVisibleDescendants(state, index);

      if (isNodeExpanded(updatedNode)) {
        const updatedNodeSubTree = getFlattenedTreePaths([updatedNode], parents);

        flattenedTree.splice(index + 1, 0, ...updatedNodeSubTree.slice(1));
      } else {
        flattenedTree.splice(index + 1, numberOfVisibleDescendants);
      }
    }

    const tree = replaceNodeFromTree(state.tree, {...updatedNode, parents});

    return new State(tree, flattenedTree);
  };

  /**
   * Given a state, deletes a node
   * @param {State} state - The current state
   * @param {number} index - The visible row index
   * @return {State} An internal state representation
   */
  static deleteNodeAt = (state, index) => {
    const node = TreeState.getNodeAt(state, index);

    const flattenedTree = [...state.flattenedTree];
    const flattenedNodeMap = flattenedTree[index];
    const parents = flattenedNodeMap.slice(0, flattenedNodeMap.length - 1);

    const numberOfVisibleDescendants = nodeHasChildren(node)
      ? TreeState.getNumberOfVisibleDescendants(state, index)
      : 0;

    flattenedTree.splice(index, 1 + numberOfVisibleDescendants);

    const tree = deleteNodeFromTree(state.tree, {...node, parents});

    return new State(tree, flattenedTree);
  };
}
