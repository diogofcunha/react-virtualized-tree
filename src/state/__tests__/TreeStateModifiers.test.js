import deepFreeze from 'deep-freeze';
import {diff} from 'deep-diff';

import TreeStateModifiers from '../TreeStateModifiers';
import {Nodes} from '../../../testData/sampleTree';
import TreeState from '../TreeState';

describe('TreeStateModifiers', () => {
  const noop = () => {};

  describe('editNodeAt', () => {
    test('should fail when invalid state is supplied', () => {
      expect(() => TreeStateModifiers.editNodeAt('state', 0, noop)).toThrowError(
        'Expected a State instance but got string',
      );
      expect(() => TreeStateModifiers.editNodeAt(1225, 0, noop)).toThrowError(
        'Expected a State instance but got number',
      );
      expect(() => TreeStateModifiers.editNodeAt([], 0, noop)).toThrowError('Expected a State instance but got object');
      expect(() => TreeStateModifiers.editNodeAt({}, 0, noop)).toThrowError('Expected a State instance but got object');
      expect(() => TreeStateModifiers.editNodeAt(true, 0, noop)).toThrowError(
        'Expected a State instance but got boolean',
      );
      expect(() => TreeStateModifiers.editNodeAt(() => {}, 0, noop)).toThrowError(
        'Expected a State instance but got function',
      );
    });

    test('should fail with descriptive error when node at index does not exist', () => {
      expect(() =>
        TreeStateModifiers.editNodeAt(TreeState.createFromTree(Nodes), 20, noop),
      ).toThrowErrorMatchingSnapshot();
    });

    describe('flattened tree', () => {
      test('should collapse a node in a root node', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.editNodeAt(state, 0, n => ({
          ...n,
          state: {...n.state, expanded: false},
        }));

        expect(flattenedTree).toMatchSnapshot();
      });

      test('should collapse a node in a children node', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.editNodeAt(state, 1, n => ({
          ...n,
          state: {...n.state, expanded: false},
        }));

        expect(flattenedTree).toMatchSnapshot();
      });

      test('should expand a node in a root node', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.editNodeAt(state, 5, n => ({
          ...n,
          state: {...n.state, expanded: true},
        }));

        expect(flattenedTree).toMatchSnapshot();
      });

      test('should expand a node in a children node', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.editNodeAt(state, 2, n => ({
          ...n,
          state: {...n.state, expanded: true},
        }));

        expect(flattenedTree).toMatchSnapshot();
      });

      test('should not change for updates that do not change state', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.editNodeAt(state, 2, n => ({
          ...n,
          name: 'node',
        }));

        expect(flattenedTree).toEqual(state.flattenedTree);
      });

      test('should not change for updates that change state but not expansion', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.editNodeAt(state, 2, n => ({
          ...n,
          state: {...n.state, favorite: true},
        }));

        expect(flattenedTree).toEqual(state.flattenedTree);

        const {flattenedTree: flattenedTree2} = TreeStateModifiers.editNodeAt(state, 0, n => ({
          ...n,
          state: {...n.state, deletable: true},
        }));

        expect(flattenedTree2).toEqual(state.flattenedTree);

        const {flattenedTree: flattenedTree3} = TreeStateModifiers.editNodeAt(state, 0, n => ({
          ...n,
          state: {...n.state, randomKey: true},
        }));

        expect(flattenedTree3).toEqual(state.flattenedTree);
      });
    });

    describe('tree', () => {
      test('should update a node in the root and keep the rest intact', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const updatedName = 'Edit node 1';

        // Change 'Leaf 1'
        const {tree} = TreeStateModifiers.editNodeAt(state, 0, n => ({
          ...n,
          name: updatedName,
        }));

        const changes = diff(state.tree, tree);

        expect(changes.length).toBe(1);
        expect(changes[0]).toMatchSnapshot();
      });

      test('should update a child node and keep the rest intact', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const updatedName = 'Edited node';

        // Change 'Leaf 3'
        const {tree} = TreeStateModifiers.editNodeAt(state, 2, n => ({
          ...n,
          name: updatedName,
        }));

        const changes = diff(state.tree, tree);

        expect(changes.length).toBe(1);
        expect(changes[0]).toMatchSnapshot();
      });

      test('should update a node state in the root and keep the rest intact', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        // Expand 'Leaf 6'
        const {tree} = TreeStateModifiers.editNodeAt(state, 5, n => ({
          ...n,
          state: {expanded: true},
        }));

        const changes = diff(state.tree, tree);

        expect(changes).toMatchSnapshot();
      });

      test('should update a child node state and keep the rest intact', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        // Collapse 'Leaf 2'
        const {tree} = TreeStateModifiers.editNodeAt(state, 1, n => ({
          ...n,
          state: {...n.state, expanded: false},
        }));

        const changes = diff(state.tree, tree);

        expect(changes.length).toBe(1);
        expect(changes[0]).toMatchSnapshot();
      });

      test('should create state for a child node and keep the rest intact', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        // Favorite 'Leaf 5'
        const {tree} = TreeStateModifiers.editNodeAt(state, 4, n => ({
          ...n,
          state: {...n.state, expanded: true},
        }));

        const changes = diff(state.tree, tree);

        expect(changes.length).toBe(1);
        expect(changes[0]).toMatchSnapshot();
      });

      test('should delete state for a root node and keep the rest intact', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        // Clear state for 'Leaf 6'
        const {tree} = TreeStateModifiers.editNodeAt(state, 5, n => {
          return Object.keys(n)
            .filter(k => k !== 'state')
            .reduce((node, k) => ({...node, [k]: n[k]}), {});
        });

        const changes = diff(state.tree, tree);

        expect(changes.length).toBe(1);
        expect(changes[0]).toMatchSnapshot();
      });

      test('should delete state for a child node and keep the rest intact', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        // Clear state for 'Leaf 3'
        const {tree} = TreeStateModifiers.editNodeAt(state, 2, n => {
          return Object.keys(n)
            .filter(k => k !== 'state')
            .reduce((node, k) => ({...node, [k]: n[k]}), {});
        });

        const changes = diff(state.tree, tree);

        expect(changes.length).toBe(1);
        expect(changes[0]).toMatchSnapshot();
      });
    });
  });

  describe('deleteNodeAt', () => {
    test('should fail when invalid state is supplied', () => {
      expect(() => TreeStateModifiers.deleteNodeAt('state', 0)).toThrowError(
        'Expected a State instance but got string',
      );
      expect(() => TreeStateModifiers.deleteNodeAt(1225, 0)).toThrowError('Expected a State instance but got number');
      expect(() => TreeStateModifiers.deleteNodeAt([], 0)).toThrowError('Expected a State instance but got object');
      expect(() => TreeStateModifiers.deleteNodeAt({}, 0)).toThrowError('Expected a State instance but got object');
      expect(() => TreeStateModifiers.deleteNodeAt(true, 0)).toThrowError('Expected a State instance but got boolean');
      expect(() => TreeStateModifiers.deleteNodeAt(() => {}, 0)).toThrowError(
        'Expected a State instance but got function',
      );
    });

    test('should fail with descriptive error when node at index does not exist', () => {
      expect(() => TreeStateModifiers.deleteNodeAt(TreeState.createFromTree(Nodes), 20)).toThrowErrorMatchingSnapshot();
    });

    describe('flattened tree', () => {
      test('should delete a root node with expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.deleteNodeAt(state, 0);

        expect(flattenedTree).toMatchSnapshot();
      });

      test('should delete a root node without expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.deleteNodeAt(state, 5);

        expect(flattenedTree).toMatchSnapshot();
      });

      test('should delete a child node with expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.deleteNodeAt(state, 1);

        expect(flattenedTree).toMatchSnapshot();
      });

      test('should delete a child node without expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {flattenedTree} = TreeStateModifiers.deleteNodeAt(state, 2);

        expect(flattenedTree).toMatchSnapshot();
      });
    });

    describe('tree', () => {
      test('should delete a root node with expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {tree} = TreeStateModifiers.deleteNodeAt(state, 0);

        const changes = diff(state.tree, tree);

        expect(changes).toMatchSnapshot();
      });

      test('should delete a root node without expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {tree} = TreeStateModifiers.deleteNodeAt(state, 6);

        const changes = diff(state.tree, tree);

        expect(changes).toMatchSnapshot();
      });

      test('should delete a child node without expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {tree} = TreeStateModifiers.deleteNodeAt(state, 2);

        const changes = diff(state.tree, tree);

        expect(changes).toMatchSnapshot();
      });

      test('should delete a child node with expanded children', () => {
        const state = TreeState.createFromTree(Nodes);

        deepFreeze(state);

        const {tree} = TreeStateModifiers.deleteNodeAt(state, 1);

        const changes = diff(state.tree, tree);

        expect(changes).toMatchSnapshot();
      });
    });
  });
});
