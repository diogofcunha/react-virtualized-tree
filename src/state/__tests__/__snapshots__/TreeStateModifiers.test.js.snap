// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeStateModifiers deleteNodeAt flattened tree should delete a child node with expanded children 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers deleteNodeAt flattened tree should delete a child node without expanded children 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    2,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers deleteNodeAt flattened tree should delete a root node with expanded children 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers deleteNodeAt flattened tree should delete a root node without expanded children 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    2,
    3,
  ],
  Array [
    0,
    2,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers deleteNodeAt should fail with descriptive error when node at index does not exist 1`] = `"Tried to get node at row \\"20\\" but got nothing, the tree are 7 visible rows"`;

exports[`TreeStateModifiers deleteNodeAt tree should delete a child node with expanded children 1`] = `
Array [
  DiffArray {
    "index": 1,
    "item": DiffDeleted {
      "kind": "D",
      "lhs": Object {
        "id": 5,
        "name": "Leaf 5",
      },
    },
    "kind": "A",
    "path": Array [
      0,
      "children",
    ],
  },
  DiffEdit {
    "kind": "E",
    "lhs": 2,
    "path": Array [
      0,
      "children",
      0,
      "id",
    ],
    "rhs": 5,
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 2",
    "path": Array [
      0,
      "children",
      0,
      "name",
    ],
    "rhs": "Leaf 5",
  },
  DiffDeleted {
    "kind": "D",
    "lhs": Object {
      "deletable": true,
      "expanded": true,
    },
    "path": Array [
      0,
      "children",
      0,
      "state",
    ],
  },
  DiffDeleted {
    "kind": "D",
    "lhs": Array [
      Object {
        "children": Array [
          Object {
            "id": "c-3",
            "name": "Leaf 3 Child",
            "state": Object {},
          },
        ],
        "id": 3,
        "name": "Leaf 3",
        "state": Object {
          "deletable": true,
          "expanded": false,
          "favorite": true,
        },
      },
      Object {
        "id": 4,
        "name": "Leaf 4",
      },
    ],
    "path": Array [
      0,
      "children",
      0,
      "children",
    ],
  },
]
`;

exports[`TreeStateModifiers deleteNodeAt tree should delete a child node without expanded children 1`] = `
Array [
  DiffArray {
    "index": 1,
    "item": DiffDeleted {
      "kind": "D",
      "lhs": Object {
        "id": 4,
        "name": "Leaf 4",
      },
    },
    "kind": "A",
    "path": Array [
      0,
      "children",
      0,
      "children",
    ],
  },
  DiffEdit {
    "kind": "E",
    "lhs": 3,
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "id",
    ],
    "rhs": 4,
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 3",
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "name",
    ],
    "rhs": "Leaf 4",
  },
  DiffDeleted {
    "kind": "D",
    "lhs": Object {
      "deletable": true,
      "expanded": false,
      "favorite": true,
    },
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "state",
    ],
  },
  DiffDeleted {
    "kind": "D",
    "lhs": Array [
      Object {
        "id": "c-3",
        "name": "Leaf 3 Child",
        "state": Object {},
      },
    ],
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "children",
    ],
  },
]
`;

exports[`TreeStateModifiers deleteNodeAt tree should delete a root node with expanded children 1`] = `
Array [
  DiffArray {
    "index": 2,
    "item": DiffDeleted {
      "kind": "D",
      "lhs": Object {
        "id": "z",
        "name": "Leaf z",
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
    },
    "kind": "A",
  },
  DiffEdit {
    "kind": "E",
    "lhs": 1,
    "path": Array [
      1,
      "id",
    ],
    "rhs": "z",
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 6",
    "path": Array [
      1,
      "name",
    ],
    "rhs": "Leaf z",
  },
  DiffDeleted {
    "kind": "D",
    "lhs": false,
    "path": Array [
      1,
      "state",
      "expanded",
    ],
  },
  DiffNew {
    "kind": "N",
    "path": Array [
      1,
      "state",
      "favorite",
    ],
    "rhs": true,
  },
  DiffDeleted {
    "kind": "D",
    "lhs": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "path": Array [
      1,
      "children",
    ],
  },
  DiffEdit {
    "kind": "E",
    "lhs": 0,
    "path": Array [
      0,
      "id",
    ],
    "rhs": 1,
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 1",
    "path": Array [
      0,
      "name",
    ],
    "rhs": "Leaf 6",
  },
  DiffEdit {
    "kind": "E",
    "lhs": true,
    "path": Array [
      0,
      "state",
      "expanded",
    ],
    "rhs": false,
  },
  DiffNew {
    "kind": "N",
    "path": Array [
      0,
      "state",
      "deletable",
    ],
    "rhs": true,
  },
  DiffEdit {
    "kind": "E",
    "lhs": 5,
    "path": Array [
      0,
      "children",
      1,
      "id",
    ],
    "rhs": 9,
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 5",
    "path": Array [
      0,
      "children",
      1,
      "name",
    ],
    "rhs": "Leaf 10",
  },
  DiffEdit {
    "kind": "E",
    "lhs": 2,
    "path": Array [
      0,
      "children",
      0,
      "id",
    ],
    "rhs": 6,
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 2",
    "path": Array [
      0,
      "children",
      0,
      "name",
    ],
    "rhs": "Leaf 7",
  },
  DiffEdit {
    "kind": "E",
    "lhs": true,
    "path": Array [
      0,
      "children",
      0,
      "state",
      "expanded",
    ],
    "rhs": false,
  },
  DiffDeleted {
    "kind": "D",
    "lhs": true,
    "path": Array [
      0,
      "children",
      0,
      "state",
      "deletable",
    ],
  },
  DiffEdit {
    "kind": "E",
    "lhs": 4,
    "path": Array [
      0,
      "children",
      0,
      "children",
      1,
      "id",
    ],
    "rhs": 8,
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 4",
    "path": Array [
      0,
      "children",
      0,
      "children",
      1,
      "name",
    ],
    "rhs": "Leaf 9",
  },
  DiffEdit {
    "kind": "E",
    "lhs": 3,
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "id",
    ],
    "rhs": 7,
  },
  DiffEdit {
    "kind": "E",
    "lhs": "Leaf 3",
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "name",
    ],
    "rhs": "Leaf 8",
  },
  DiffDeleted {
    "kind": "D",
    "lhs": Object {
      "deletable": true,
      "expanded": false,
      "favorite": true,
    },
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "state",
    ],
  },
  DiffDeleted {
    "kind": "D",
    "lhs": Array [
      Object {
        "id": "c-3",
        "name": "Leaf 3 Child",
        "state": Object {},
      },
    ],
    "path": Array [
      0,
      "children",
      0,
      "children",
      0,
      "children",
    ],
  },
]
`;

exports[`TreeStateModifiers deleteNodeAt tree should delete a root node without expanded children 1`] = `
Array [
  DiffArray {
    "index": 2,
    "item": DiffDeleted {
      "kind": "D",
      "lhs": Object {
        "id": "z",
        "name": "Leaf z",
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
    },
    "kind": "A",
  },
]
`;

exports[`TreeStateModifiers editNodeAt flattened tree should collapse a node in a children node 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers editNodeAt flattened tree should collapse a node in a root node 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers editNodeAt flattened tree should expand a node in a children node 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    2,
    3,
  ],
  Array [
    0,
    2,
    3,
    "c-3",
  ],
  Array [
    0,
    2,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers editNodeAt flattened tree should expand a node in a root node 1`] = `
Array [
  Array [
    0,
  ],
  Array [
    0,
    2,
  ],
  Array [
    0,
    2,
    3,
  ],
  Array [
    0,
    2,
    4,
  ],
  Array [
    0,
    5,
  ],
  Array [
    1,
  ],
  Array [
    1,
    6,
  ],
  Array [
    1,
    9,
  ],
  Array [
    "z",
  ],
]
`;

exports[`TreeStateModifiers editNodeAt should fail with descriptive error when node at index does not exist 1`] = `"Tried to get node at row \\"20\\" but got nothing, the tree are 7 visible rows"`;

exports[`TreeStateModifiers editNodeAt tree should create state for a child node and keep the rest intact 1`] = `
DiffNew {
  "kind": "N",
  "path": Array [
    0,
    "children",
    1,
    "state",
  ],
  "rhs": Object {
    "expanded": true,
  },
}
`;

exports[`TreeStateModifiers editNodeAt tree should delete state for a child node and keep the rest intact 1`] = `
DiffDeleted {
  "kind": "D",
  "lhs": Object {
    "deletable": true,
    "expanded": false,
    "favorite": true,
  },
  "path": Array [
    0,
    "children",
    0,
    "children",
    0,
    "state",
  ],
}
`;

exports[`TreeStateModifiers editNodeAt tree should delete state for a root node and keep the rest intact 1`] = `
DiffDeleted {
  "kind": "D",
  "lhs": Object {
    "deletable": true,
    "expanded": false,
  },
  "path": Array [
    1,
    "state",
  ],
}
`;

exports[`TreeStateModifiers editNodeAt tree should update a child node and keep the rest intact 1`] = `
DiffEdit {
  "kind": "E",
  "lhs": "Leaf 3",
  "path": Array [
    0,
    "children",
    0,
    "children",
    0,
    "name",
  ],
  "rhs": "Edited node",
}
`;

exports[`TreeStateModifiers editNodeAt tree should update a child node state and keep the rest intact 1`] = `
DiffEdit {
  "kind": "E",
  "lhs": true,
  "path": Array [
    0,
    "children",
    0,
    "state",
    "expanded",
  ],
  "rhs": false,
}
`;

exports[`TreeStateModifiers editNodeAt tree should update a node in the root and keep the rest intact 1`] = `
DiffEdit {
  "kind": "E",
  "lhs": "Leaf 1",
  "path": Array [
    0,
    "name",
  ],
  "rhs": "Edit node 1",
}
`;

exports[`TreeStateModifiers editNodeAt tree should update a node state in the root and keep the rest intact 1`] = `
Array [
  DiffEdit {
    "kind": "E",
    "lhs": false,
    "path": Array [
      1,
      "state",
      "expanded",
    ],
    "rhs": true,
  },
  DiffDeleted {
    "kind": "D",
    "lhs": true,
    "path": Array [
      1,
      "state",
      "deletable",
    ],
  },
]
`;
